File: app/src/main/java/com/king/mvvmframe/app/dictionary/DictionaryActivity.java
package com.king.mvvmframe.app.dictionary;

import android.os.Bundle;
import android.text.InputFilter;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.king.base.adapter.divider.DividerItemDecoration;
import com.king.base.util.StringUtils;
import com.king.base.util.ToastUtils;
import com.king.frame.mvvmframe.base.BaseActivity;
import com.king.frame.mvvmframe.base.livedata.StatusEvent;
import com.king.mvvmframe.R;
import com.king.mvvmframe.app.adapter.BindingAdapter;
import com.king.mvvmframe.app.adapter.SearchHistoryAdapter;
import com.king.mvvmframe.bean.SearchHistory;
import com.king.mvvmframe.databinding.DictionaryActivityBinding;


import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.LinearLayoutManager;
import dagger.hilt.android.AndroidEntryPoint;
import timber.log.Timber;

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
@AndroidEntryPoint
public class DictionaryActivity extends BaseActivity<DictionaryViewModel, DictionaryActivityBinding> {

    private SearchView searchView;

    private BindingAdapter<String> mAdapter;

    private SearchHistoryAdapter mSearchHistoryAdapter;

    private List<SearchHistory> listHistory;

    private String word;

    @Override
    public int getLayoutId() {
        return R.layout.dictionary_activity;
    }

    @Override
    public void initData(@Nullable Bundle savedInstanceState) {

        getViewDataBinding().recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        getViewDataBinding().recyclerView.addItemDecoration(new DividerItemDecoration(getContext(),DividerItemDecoration.VERTICAL));

        mAdapter = new BindingAdapter<>(getContext(),R.layout.rv_dictionary_item);

        getViewDataBinding().recyclerView.setAdapter(mAdapter);

        listHistory = new ArrayList<>();
        mSearchHistoryAdapter = new SearchHistoryAdapter(getContext(),listHistory);
        getViewDataBinding().tflHistory.setAdapter(mSearchHistoryAdapter);

        getViewDataBinding().tflHistory.setOnTagClickListener((view, position, parent) -> {
            search(mSearchHistoryAdapter.getItem(position).getName());
            return true;
        });

        getViewModel().getDictionaryLiveData().observe(this, data -> {
            mAdapter.refreshData(data.getXiangjie());
            searchView.clearFocus();
        });

        registerStatusEvent(status -> {
            switch (status){
                case StatusEvent.Status.LOADING:
                    if(!getViewDataBinding().srl.isRefreshing()){
                        showLoading();
                    }
                    break;
                case StatusEvent.Status.SUCCESS:
                case StatusEvent.Status.FAILURE:
                case StatusEvent.Status.ERROR:
                    hideLoading();
                    getViewDataBinding().srl.setRefreshing(false);
                    break;
            }
        });

        registerMessageEvent(message -> {
            Timber.d("message:%s" , message);
            ToastUtils.showToast(getContext(), message);
        });

        getViewModel().getSearchHistoryLiveData().observe(this,list ->{
            refreshSearchHistory(list);
        });
        getViewDataBinding().srl.setOnRefreshListener(()-> {
            search(word);
        });

    }

    private void search(String key){
        if(StringUtils.isNotBlank(key)){
            searchView.setQuery(key,true);
        }else{
            getViewDataBinding().srl.setRefreshing(false);
        }

    }

    private void refreshSearchHistory(List<SearchHistory> data){
        listHistory.clear();
        if(data!=null && data.size() > 0){
            listHistory.addAll(data);
            getViewDataBinding().ivDeleteHistory.setVisibility(View.VISIBLE);
        }else{
            getViewDataBinding().ivDeleteHistory.setVisibility(View.INVISIBLE);
        }
        mSearchHistoryAdapter.notifyDataChanged();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.search_view_menu, menu);
        MenuItem searchItem = menu.findItem(R.id.searchItem);
        //通过MenuItem得到SearchView
        searchView = (SearchView)searchItem.getActionView();
        SearchView.SearchAutoComplete searchSrcTextView = searchView.findViewById(R.id.search_src_text);
        if(searchSrcTextView != null){
            searchSrcTextView.setFilters(new InputFilter[]{new InputFilter.LengthFilter(1)});
        }
        searchView.setQueryHint("输入需要查询的汉字");
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                if(StringUtils.isNotBlank(s)){
                    word = s;
                    getViewModel().getDictionaryInfo(word);
                }else{
                    ToastUtils.showToast(getContext(),R.string.tips_search_content);
                }
                return true;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                return false;
            }
        });
        return super.onCreateOptionsMenu(menu);
    }

    private void clickDelete(){
        ToastUtils.showToast(getContext(),"点击清空");
        getViewModel().deleteAllHistory();

    }

    public void onClick(View v){
        switch (v.getId()){
            case R.id.ivDeleteHistory:
                clickDelete();
                break;
        }
    }
}

File: app/src/main/java/com/king/mvvmframe/app/dictionary/DictionaryActivity.kt
package com.king.mvvmframe.app.dictionary

import android.annotation.SuppressLint
import android.os.Bundle
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.view.Menu
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.SearchView.SearchAutoComplete
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.king.base.adapter.divider.DividerItemDecoration
import com.king.base.util.StringUtils
import com.king.frame.mvvmframe.base.BaseActivity
import com.king.mvvmframe.R
import com.king.mvvmframe.app.adapter.BindingAdapter
import com.king.mvvmframe.app.adapter.SearchHistoryAdapter
import com.king.mvvmframe.data.model.SearchHistory
import com.king.mvvmframe.databinding.DictionaryActivityBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

/**
 * 示例
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
@AndroidEntryPoint
class DictionaryActivity : BaseActivity<DictionaryViewModel, DictionaryActivityBinding>() {

    private var searchView: SearchView? = null

    private val mAdapter by lazy {
        BindingAdapter<String>(getContext(), R.layout.rv_dictionary_item)
    }

    private val listHistory = mutableListOf<SearchHistory>()

    private val mSearchHistoryAdapter by lazy {
        SearchHistoryAdapter(listHistory)
    }

    private var word: String = ""

    override fun getLayoutId(): Int {
        return R.layout.dictionary_activity
    }

    override fun initData(savedInstanceState: Bundle?) {
        binding.recyclerView.layoutManager = LinearLayoutManager(getContext())
        binding.recyclerView.addItemDecoration(
            DividerItemDecoration(
                getContext(),
                DividerItemDecoration.VERTICAL
            )
        )
        binding.recyclerView.adapter = mAdapter

        binding.tflHistory.adapter = mSearchHistoryAdapter

        binding.tflHistory.setOnTagClickListener { _, position, _ ->
            search(mSearchHistoryAdapter.getItem(position).word)
            true
        }

        binding.srl.setOnRefreshListener {
            search(word)
        }

        lifecycleScope.launch {
            viewModel.dictionaryFlow.flowWithLifecycle(lifecycle).collect {
                mAdapter.refreshData(it?.xiangjie)
            }
        }

        lifecycleScope.launch {
            viewModel.getSearchHistoryFlow().flowWithLifecycle(lifecycle).collect {
                listHistory.clear()
                listHistory.addAll(it)
                mSearchHistoryAdapter.notifyDataChanged()
            }
        }
    }

    private fun search(key: String) {
        if (StringUtils.isNotBlank(key)) {
            searchView?.setQuery(key, true)
        } else {
            binding.srl.isRefreshing = false
        }
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.search_view_menu, menu)
        val searchItem = menu.findItem(R.id.searchItem)
        //通过MenuItem得到SearchView
        searchView = searchItem.actionView as SearchView?
        val searchSrcTextView = searchView?.findViewById<SearchAutoComplete>(R.id.search_src_text)
        searchSrcTextView?.also {
            it.filters = arrayOf<InputFilter>(LengthFilter(1))
        }
        searchView?.queryHint = getString(R.string.word_search_hint)
        searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                if (StringUtils.isNotBlank(s)) {
                    word = s
                    viewModel.getDictionaryInfo(word)
                    searchView?.clearFocus()
                } else {
                    showToast(getString(R.string.tips_search_content))
                }
                return true
            }

            override fun onQueryTextChange(s: String): Boolean {
                return false
            }
        })
        return super.onCreateOptionsMenu(menu)
    }

    override fun showLoading() {
        if (!binding.srl.isRefreshing) {
            super.showLoading()
        }
    }

    override fun hideLoading() {
        super.hideLoading()
        binding.srl.isRefreshing = false
    }

    private fun clearSearchHistory() {
        viewModel.clearSearchHistory()
        mAdapter.refreshData(null)
        word = ""
        searchView?.setQuery(word, false)
    }

    fun onClick(v: View) {
        when (v.id) {
            R.id.ivDeleteHistory -> clearSearchHistory()
        }
    }
}
